components/gitpod-db/src/typeorm/user-db-impl.ts

```diff
@@ -395,6 +402,43 @@ export class TypeORMUserDBImpl implements UserDB {
         await repo.save(envVar);
     }
 
+    public async hasSSHPublicKey(userId: string): Promise<boolean> {
+        const repo = await this.getSSHPublicKeyRepo();
+        return !!(await repo.findOne({ where: { userId, deleted: false } }));
+    }
+
+    public async getSSHPublicKeys(userId: string): Promise<UserSSHPublicKey[]> {
+        const repo = await this.getSSHPublicKeyRepo();
+        return repo.find({ where: { userId, deleted: false }, order: { creationTime: "ASC" } });
+    }
+
+    public async addSSHPublicKey(userId: string, value: SSHPublicKeyValue): Promise<UserSSHPublicKey> {
+        const repo = await this.getSSHPublicKeyRepo();
+        const fingerprint = SSHPublicKeyValue.getFingerprint(value);
+        const allKeys = await repo.find({ where: { userId, deleted: false } });
+        const prevOne = allKeys.find((e) => e.fingerprint === fingerprint);
+        if (!!prevOne) {
+            throw new Error(`Duplicate public key with ${prevOne.name}`);
+        }
+        if (allKeys.length > SSHPublicKeyValue.MAXIMUM_KEY_LENGTH) {
+            throw new Error(`The maximum of public keys is ${SSHPublicKeyValue.MAXIMUM_KEY_LENGTH}`);
+        }
+        return repo.save({
+            id: uuidv4(),
+            userId,
+            fingerprint,
+            name: value.name,
+            key: value.key,
+            creationTime: new Date().toISOString(),
+            deleted: false,
+        });
+    }
+
+    public async deleteSSHPublicKey(userId: string, id: string): Promise<void> {
+        const repo = await this.getSSHPublicKeyRepo();
+        await repo.update({ userId, id }, { deleted: true });
```

Comment:
This won't necessarily hit the PK index. Why not use `{ id }` alone? That should guarantee this to work as fast as possible. :+1:

Reply:
To make sure this public key belongs to the user

Reply:
> To make sure this public key belongs to the user

Ah, got it. :+1: